prometheus_juju_exporter/collector.py

```diff
@@ -140,6 +140,33 @@ def _get_machine_type(self, machine: Dict) -> MachineType:
 
         return MachineType.METAL
 
+    def _get_machine_sensible_hostname(self, machine: Dict) -> str:
+        """Try to find a sensible hostname for the machine.
+
+        The hostname field is only supported for juju controller versions
+        2.8.10 and upwards. For the remaining lower versions, this field
+        exists however is None or "None". This method tries to get it from
+        hostname if it can, else it will try to get it from the "instance-id"
+        field, which seems to exist almost always.
```

Comment:
If the `instance-id` is not a foolproof fallback, I think we should add some handling for when machine does not have it. At least a log message to help identify that there's an issue.

Reply:
hmm yeah i think it is a good idea. I am thinking of a log message with level ERROR to indicate this. or warning perhaps?

Reply:
yup, error level seems appropriate. For the future we should get to exporting errors as metrics (Do we have task for tracking this @agileshaw ?).

Reply:
By the way, I just remembered that there's another scenario where `hostname` could be empty: the machine instance is created but not ready yet. I'm not sure whether `instance-id` would appear in this case. It definitely worth some thorough testing.

Reply:
came across this while testing: container returned `None` hostname, and `pending` as `instance-id`. I think we should add `pending` as the machine identifiers we do not like alongside with `None, "None"` so that we don't end up with a machine with key `pending`

Reply:
can't we run `juju run --machine 0 hostname`

Reply:
In the config file, I am using the credentials of a juju user who only has `read` level access to the models on a controller. ssh'ing requires a higher access level granted. I think we should not enforce write level access from a security point of view

Reply:
+1 on Martin's comment. And since we are running the snap inside a charm, running juju commands is not feasible.